File: 14-TypeScript/Grey Multi-Tenant/packages/grey-php/src/Error/Result.php
<?php

declare(strict_types=1);

namespace Grey\Sdk\Error;

/**
 * Represents the result of an operation that can succeed or fail.
 *
 * @template T
 */
final class Result
{
    /**
     * @param T|null $data
     */
    private function __construct(
        private readonly bool $ok,
        private readonly mixed $data = null,
        private readonly ?GreyError $error = null,
    ) {}

    /**
     * Creates a successful result.
     *
     * @template U
     * @param U $data
     * @return Result<U>
     */
    public static function ok(mixed $data): self
    {
        return new self(true, $data);
    }

    /**
     * Creates a failed result.
     *
     * @return Result<null>
     */
    public static function err(GreyError $error): self
    {
        return new self(false, null, $error);
    }

    /**
     * Returns true if the result is successful.
     */
    public function isOk(): bool
    {
        return $this->ok;
    }

    /**
     * Returns true if the result is an error.
     */
    public function isErr(): bool
    {
        return !$this->ok;
    }

    /**
     * Gets the data if successful, null otherwise.
     *
     * @return T|null
     */
    public function getData(): mixed
    {
        return $this->data;
    }

    /**
     * Gets the error if failed, null otherwise.
     */
    public function getError(): ?GreyError
    {
        return $this->error;
    }

    /**
     * Gets the data or throws an exception if failed.
     *
     * @return T
     * @throws \RuntimeException
     */
    public function unwrap(): mixed
    {
        if (!$this->ok) {
            throw new \RuntimeException(
                "Unwrap failed: " . ($this->error?->getMessage() ?? 'Unknown error')
            );
        }
        return $this->data;
    }

    /**
     * Gets the data or returns a default value if failed.
     *
     * @template U
     * @param U $default
     * @return T|U
     */
    public function unwrapOr(mixed $default): mixed
    {
        return $this->ok ? $this->data : $default;
    }

    /**
     * Maps the data if successful.
     *
     * @template U
     * @param callable(T): U $fn
     * @return Result<U>
     */
    public function map(callable $fn): self
    {
        if ($this->ok) {
            try {
                return self::ok($fn($this->data));
            } catch (\Throwable $e) {
                return self::err(GreyError::fromException($e));
            }
        }
        return $this;
    }

    /**
     * Chains another operation if successful.
     *
     * @template U
     * @param callable(T): Result<U> $fn
     * @return Result<U>
     */
    public function then(callable $fn): self
    {
        if ($this->ok) {
            try {
                return $fn($this->data);
            } catch (\Throwable $e) {
                return self::err(GreyError::fromException($e));
            }
        }
        return $this;
    }

    /**
     * Handles an error if failed.
     *
     * @param callable(GreyError): Result<T> $fn
     * @return Result<T>
     */
    public function catch(callable $fn): self
    {
        if (!$this->ok && $this->error !== null) {
            try {
                return $fn($this->error);
            } catch (\Throwable $e) {
                return self::err(GreyError::fromException($e));
            }
        }
        return $this;
    }

    /**
     * Converts the result to an array.
     */
    public function toArray(): array
    {
        if ($this->ok) {
            return ['ok' => true, 'data' => $this->data];
        }
        return ['ok' => false, 'error' => $this->error?->toArray()];
    }
}

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-ruby/lib/grey_sdk/error/result.rb
# frozen_string_literal: true

module GreySdk
  module Error
    # Represents the result of an operation that can succeed or fail
    class Result
      attr_reader :data, :error

      # @param ok [Boolean] Whether the result is successful
      # @param data [Object, nil] The success data
      # @param error [GreyError, nil] The error
      def initialize(ok:, data: nil, error: nil)
        @ok = ok
        @data = data
        @error = error
      end

      # Creates a successful result
      # @param data [Object] The success data
      # @return [Result]
      def self.ok(data)
        new(ok: true, data: data)
      end

      # Creates a failed result
      # @param error [GreyError] The error
      # @return [Result]
      def self.err(error)
        error = GreyError.new(ErrorCode::UNKNOWN, error.to_s) unless error.is_a?(GreyError)
        new(ok: false, error: error)
      end

      # Returns true if the result is successful
      # @return [Boolean]
      def ok?
        @ok
      end

      # Returns true if the result is an error
      # @return [Boolean]
      def err?
        !@ok
      end

      # Gets the data or raises an exception if failed
      # @return [Object]
      # @raise [RuntimeError] If the result is an error
      def unwrap!
        raise "Unwrap failed: #{@error&.message || 'Unknown error'}" unless ok?

        @data
      end

      # Gets the data or returns a default value if failed
      # @param default [Object] Default value
      # @return [Object]
      def unwrap_or(default)
        ok? ? @data : default
      end

      # Maps the data if successful
      # @yield [data] Block to transform the data
      # @return [Result]
      def map(&block)
        return self unless ok?

        begin
          Result.ok(block.call(@data))
        rescue StandardError => e
          Result.err(GreyError.from_exception(e))
        end
      end

      # Chains another operation if successful
      # @yield [data] Block that returns a Result
      # @return [Result]
      def then(&block)
        return self unless ok?

        begin
          block.call(@data)
        rescue StandardError => e
          Result.err(GreyError.from_exception(e))
        end
      end

      # Handles an error if failed
      # @yield [error] Block that returns a Result
      # @return [Result]
      def catch(&block)
        return self if ok?

        begin
          block.call(@error)
        rescue StandardError => e
          Result.err(GreyError.from_exception(e))
        end
      end

      # Converts the result to a hash
      # @return [Hash]
      def to_h
        if ok?
          { ok: true, data: @data }
        else
          { ok: false, error: @error&.to_h }
        end
      end

      # String representation
      # @return [String]
      def to_s
        if ok?
          "Result::Ok(#{@data.inspect})"
        else
          "Result::Err(#{@error})"
        end
      end

      # Inspect representation
      # @return [String]
      def inspect
        "#<GreySdk::Error::Result #{to_s}>"
      end
    end
  end
end
